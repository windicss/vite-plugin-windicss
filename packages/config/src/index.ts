import fs from 'node:fs'
import path from 'node:path'
import _debug from 'debug'
import type { JITI } from 'jiti'
import _jiti from 'jiti'
import type { FullConfig as WindiCssOptions } from 'windicss/types/interfaces'

let jiti: JITI

export type { WindiCssOptions }

export const defaultConfigureFiles = [
  'windi.config.ts',
  'windi.config.mts',
  'windi.config.js',
  'windi.config.mjs',
  'windi.config.cjs',

  'windicss.config.ts',
  'windicss.config.mts',
  'windicss.config.js',
  'windicss.config.mjs',
  'windicss.config.cjs',

  'tailwind.config.ts',
  'tailwind.config.mts',
  'tailwind.config.js',
  'tailwind.config.mjs',
  'tailwind.config.cjs',
]

export interface LoadConfigurationOptions {
  /**
   * Name for debug
   *
   * @default 'windicss-config'
   * @internal
   */
  name?: string
  /**
   * Config object or path
   */
  config?: WindiCssOptions | string
  /**
   * CWD
   *
   * @default process.cwd
   * @internal
   */
  root?: string
  /**
   * A list of filename of paths to search of config files
   */
  configFiles?: string[]
  /**
   * On loading configuration error
   *
   * @default [throw error]
   */
  onConfigurationError?: (error: unknown) => void
  /**
   * On configure file not found
   *
   * @default [emit warning]
   */
  onConfigurationNotFound?: (filepath: string) => void
}

export function loadConfiguration(options: LoadConfigurationOptions): {
  config: WindiCssOptions
  filepath?: string
  error?: unknown
} {
  if (!jiti)
    jiti = _jiti(process.cwd(), { requireCache: false, cache: false, v8cache: false })

  let resolved: WindiCssOptions = {}
  let configFilePath: string | undefined
  let error: unknown

  const {
    name = 'windicss-config',
    config,
    root = process.cwd(),
    configFiles: configureFiles = defaultConfigureFiles,
    onConfigurationError = (e) => {
      throw e
    },
    onConfigurationNotFound = (path) => {
      console.warn(`[${name}] config file "${path}" not found, ignored`)
    },
  } = options

  const debugConfig = _debug(`${name}:config`)

  if (typeof config === 'string' || !config) {
    if (!config) {
      for (const name of configureFiles) {
        const tryPath = path.resolve(root, name)
        if (fs.existsSync(tryPath)) {
          configFilePath = tryPath
          break
        }
      }
    }
    else {
      configFilePath = path.resolve(root, config)
      if (!fs.existsSync(configFilePath)) {
        onConfigurationNotFound(config)
        configFilePath = undefined
      }
    }

    if (configFilePath) {
      try {
        debugConfig('loading from ', configFilePath)

        resolved = jiti(configFilePath)
        if (resolved.default)
          resolved = resolved.default
      }
      catch (e) {
        error = e
        configFilePath = undefined
        resolved = {}
        onConfigurationError?.(e)
      }
    }
  }
  else {
    resolved = config
  }

  debugConfig(resolved)

  return {
    error,
    config: resolved,
    filepath: configFilePath,
  }
}
