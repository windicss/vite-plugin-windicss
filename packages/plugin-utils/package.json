{
  "name": "@windicss/plugin-utils",
  "version": "1.8.6",
  "description": "Common utils for building integrations of Windi CSS",
  "author": "antfu <anthonyfu117@hotmail.com>",
  "license": "MIT",
  "funding": "https://github.com/sponsors/antfu",
  "homepage": "https://github.com/antfu/vite-plugin-windicss",
  "repository": {
    "type": "git",
    "url": "https://github.com/antfu/vite-plugin-windicss"
  },
  "bugs": "https://github.com/antfu/vite-plugin-windicss/issues",
  "keywords": [
    "windicss",
    "tailwindcss"
  ],
  "exports": {
    ".": {
      "require": "./dist/index.js",
      "import": "./dist/index.mjs"
    },
    "./*": "./*"
  },
  "main": "dist/index.js",
  "module": "dist/index.mjs",
  "types": "dist/index.d.ts",
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "tsup src/index.ts --dts --format cjs,esm --no-splitting --external sucrase/register/ts,pug,windicss/utils/style,windicss/utils/parser,windicss/utils",
    "dev": "npm run build -- --watch",
    "prepublishOnly": "npm run build"
  },
  "dependencies": {
    "@antfu/utils": "^0.5.2",
    "@windicss/config": "workspace:*",
    "debug": "^4.3.4",
    "fast-glob": "^3.2.11",
    "magic-string": "^0.26.2",
    "micromatch": "^4.0.5",
    "windicss": "^3.5.6"
  },
  "devDependencies": {
    "pug": "^3.0.2"
  }
}
