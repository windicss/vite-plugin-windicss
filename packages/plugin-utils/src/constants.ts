export const defaultAlias: Record<string, TagNames> = {
  'router-link': 'a',
}

export const preflightTags = ['html', 'body', 'div']

export const htmlTags = [
  'html',
  'body',
  'div',
  'a',
  'abbr',
  'address',
  'area',
  'article',
  'aside',
  'audio',
  'base',
  'basefont',
  'bdo',
  'blink',
  'blockquote',
  'br',
  'button',
  'canvas',
  'caption',
  'center',
  'col',
  'colgroup',
  'command',
  'comment',
  'datalist',
  'dd',
  'del',
  'details',
  'dir',
  'dl',
  'dt',
  'embed',
  'fieldset',
  'figure',
  'b',
  'big',
  'i',
  'small',
  'tt',
  'font',
  'footer',
  'form',
  'frame',
  'frameset',
  'head',
  'header',
  'hgroup',
  'h1',
  'h2',
  'h3',
  'h4',
  'h5',
  'h6',
  'hr',
  'isindex',
  'iframe',
  'ilayer',
  'img',
  'input',
  'ins',
  'keygen',
  'keygen',
  'label',
  'layer',
  'legend',
  'li',
  'link',
  'map',
  'mark',
  'marquee',
  'menu',
  'meta',
  'meter',
  'multicol',
  'nav',
  'nobr',
  'noembed',
  'noframes',
  'noscript',
  'object',
  'ol',
  'optgroup',
  'option',
  'output',
  'p',
  'param',
  'cite',
  'code',
  'dfn',
  'em',
  'kbd',
  'samp',
  'strong',
  'var',
  'plaintext',
  'pre',
  'progress',
  'q',
  'ruby',
  'script',
  'section',
  'select',
  'spacer',
  'span',
  's',
  'strike',
  'style',
  'sub',
  'sup',
  'svg',
  'table',
  'tbody',
  'td',
  'textarea',
  'tfoot',
  'th',
  'thead',
  'time',
  'title',
  'tr',
  'u',
  'ul',
  'video',
  'wbr',
  'wbr',
  'xmp',
] as const

export type TagNames = (typeof htmlTags)[number]
