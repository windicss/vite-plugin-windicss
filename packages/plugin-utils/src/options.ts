import type { Extractor } from 'windicss/types/interfaces'
import type { LoadConfigurationOptions, WindiCssOptions } from '@windicss/config'
import type { TransformerFunction } from './transforms'
import type { TagNames } from './constants'
import type { WindiPluginUtils } from './createUtils'

export { WindiCssOptions }

export interface UserOptions {
  /**
   * Options for windicss/tailwindcss.
   * Also accepts string as config file path.
   *
   * @default auto searching for `windi.config.ts` / `tailwind.config.js`
   */
  config?: WindiCssOptions | string

  /**
   * A list of filename of paths to search of config files
   */
  configFiles?: string[]

  /**
   * Safe class names to be always included.
   *
   * @deprecated define this field in the windicss.config.ts instead
   */
  safelist?: string | (string | string[])[]

  /**
   * Class names to be always excluded.
   *
   * @deprecated define this field in the windicss.config.ts instead
   */
  blocklist?: string | (string | string[])[]

  /**
   * Enabled windicss preflight (a.k.a TailwindCSS style reset)
   *
   * @deprecated define this field in the windicss.config.ts instead
   * @default true
   */
  preflight?: boolean | {
    /**
     * Enable all the preflight regardless the template
     *
     * @deprecated define this field in the windicss.config.ts instead
     */
    enableAll?: boolean

    /**
     * Enable all the preflight regardless the template
     *
     * @deprecated define this field in the windicss.config.ts instead
     */
    includeAll?: boolean

    /**
     * Safelist to always included
     *
     * @deprecated define this field in the windicss.config.ts instead
     */
    safelist?: string | (string | string[])[]

    /**
     * Blocklist to always excluded
     *
     * @deprecated define this field in the windicss.config.ts instead
     */
    blocklist?: string | (string | string[])[]

    /**
      * Alias for resolving preflight
      */
    alias?: Record<string, TagNames>

    /**
     * @default true
     * @deprecated define this field in the windicss.config.ts instead
     */
    includeBase?: boolean

    /**
     * @default true
     * @deprecated define this field in the windicss.config.ts instead
     */
    includeGlobal?: boolean

    /**
     * @default true
     * @deprecated define this field in the windicss.config.ts instead
     */
    includePlugin?: boolean
  }

  /**
    * File paths will be resolved against this directory.
    *
    * @default process.cwd
    * @internal
    */
  root?: string

  /**
   * Scan the files and extract the usage
   *
   * @default true
   */
  scan?: boolean | {
    /**
     * Auto scan on startup
     *
     * @default true
     */
    runOnStartup?: boolean

    /**
     * Directories to search for classnames
     *
     * @default 'src'
     * @deprecated use `extract.include` in the windicss.config.ts instead
     */
    dirs?: string | string[]

    /**
     * File extension to search for classnames
     *
     * @default 'html', 'vue', 'md', 'mdx', 'pug', 'jsx', 'tsx', 'svelte', 'js', 'ts'
     * @deprecated use `extract.include` in the windicss.config.ts instead
     */
    fileExtensions?: string | string[]

    /**
     * Exclude globs
     *
     * @default []
     */
    exclude?: string | string[]

    /**
     * Include globs
     *
     * @default []
     */
    include?: string | string[]

    /**
     * Transformers to apply before doing extraction
     *
     * @default []
     */
    transformers?: TransformerFunction[]
  }

  /**
   * Transform CSS for `@apply` directive
   *
   * @default true
   */
  transformCSS?: boolean | 'pre' | 'post'

  /**
   * Transform groups like `hover:(bg-gray-100 font-medium)`
   *
   * @default true
   */
  transformGroups?: boolean

  /**
   * Sort the genrate utilities
   *
   * @default true
   */
  sortUtilities?: boolean

  /**
   * Callback before classes css generated
   */
  onBeforeGenerate?: (ctx: { classesPending: Set<string>; tagsPending: Set<string> }) => void

  /**
   * Callback when classes and/or tags are generated/changed
   */
  onGenerated?: (ctx: { classes: Set<string>; tags: Set<string> }) => void

  /**
   * Callback when the options are resolved. These are the plugin options and contain the windi config
   */
  onOptionsResolved?: (options: ResolvedOptions) => ResolvedOptions | void | Promise<ResolvedOptions | void>

  /**
   * Callback when the windi config is resolved. Not to be confused with the options which are the top level way to
   * configure the util package
   */
  onConfigResolved?: (config: WindiCssOptions, configFilePath?: string) => WindiCssOptions | void | Promise<WindiCssOptions | void>

  /**
   * Callback when the utils is initialized
   */
  onInitialized?: (utils: WindiPluginUtils) => void
}

export type WindiPluginUtilsOptions = Omit<LoadConfigurationOptions, 'config' | 'configFiles'> & {
  /**
   * Reuse existing plugin instance
   */
  utils?: WindiPluginUtils
}

export interface ResolvedOptions {
  config: WindiCssOptions
  configFilePath: string | undefined

  enableScan: boolean
  enablePreflight: boolean
  transformCSS: boolean | 'pre' | 'auto' | 'post'
  transformGroups: boolean

  scanOptions: {
    fileExtensions: string[]
    dirs: string[]
    exclude: string[]
    include: string[]
    runOnStartup: boolean
    transformers: TransformerFunction[]
    extractors: Extractor[]
    extraTransformTargets: {
      css: (string | ((path: string) => boolean))[]
      detect: (string | ((path: string) => boolean))[]
    }
  }

  preflightOptions: {
    includeBase: boolean
    includeGlobal: boolean
    includePlugin: boolean
    includeAll: boolean
    /**
     * @deprecated use includeAll
     */
    enableAll: boolean
    safelist: Set<string>
    blocklist: Set<string>
    alias: Record<string, string>
  }

  root: string
  sortUtilities: boolean
  safelist: Set<string>
  blocklist: Set<string>
  onBeforeGenerate: UserOptions['onBeforeGenerate']
  onGenerated: UserOptions['onGenerated']
  onConfigResolved: UserOptions['onConfigResolved']
  onOptionsResolved: UserOptions['onOptionsResolved']
  onInitialized: UserOptions['onInitialized']
}
