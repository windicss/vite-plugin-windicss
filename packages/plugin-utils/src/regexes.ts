export const regexQuotedString = /(["'`])((?:\\\1|\\\\|\n|\r|(?!<|>|\s\*\/).)*?)\1/mg
export const regexHtmlTag = /<(\w[\w-]*)([\S\s]*?)\/?[^=]>/mg
export const regexClassSplitter = /[\s'"`{}]/g
export const regexClassGroup = /([!\w+-<@][\w+:_/-]*?\w):\(((?:[!\w\s:/\\,%#.$-]|\[.*?\])*?)\)/gm
export const regexAttributifyItem = /(?:\s|^)([\w+:_/-]+)\s*=\s*(['"{])((?:\\\2|\\\\|\n|\r|.)*?)(?:\2|\})/gm

export const regexClassCheck1 = /^!?[a-z\d@<>.+-](?:\([\w,.%#\(\)+-]*\)|[\w:/\\,%#\[\].$-])*$/
export const regexClassCheck2 = /[a-z].*[\w)\]]$/
export const regexClassChecks = [
  regexClassCheck1,
  regexClassCheck2,
]

export const regexSvelteClass = /class:([\w!:/\\,%#\[\].$-]+?)=["']?\{/g

export function validClassName(i: string) {
  return regexClassChecks.every(r => i.length > 2 && i.match(r))
}
