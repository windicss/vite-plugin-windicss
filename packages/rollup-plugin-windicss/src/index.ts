import { Plugin } from 'rollup'
import _debug from 'debug'
import { createUtils, UserOptions } from '@windicss/plugin-utils'
import { createVirtualModuleLoader } from '../../shared/virtual-module'

const NAME = 'rollup-plugin-windicss'

const debug = {
  hmr: _debug(`${NAME}:hmr`),
  css: _debug(`${NAME}:transform:css`),
  group: _debug(`${NAME}:transform:group`),
  alias: _debug(`${NAME}:transform:alias`),
}

function WindiCssRollupPlugin(userOptions: UserOptions = {}): Plugin[] {
  const utils = createUtils(
    userOptions,
    {
      name: NAME,
      onConfigurationError(e) {
        throw e
      },
    },
  )
  const plugins: Plugin[] = []

  plugins.push({
    name: `${NAME}:alias`,
    async transform(code, id) {
      await utils.ensureInit()
      if (!utils.isDetectTarget(id))
        return
      debug.alias(id)
      return utils.transformAlias(code, true)
    },
  })

  if (userOptions.transformGroups !== false) {
    plugins.push({
      name: `${NAME}:groups`,
      async transform(code, id) {
        await utils.ensureInit()
        if (!utils.isDetectTarget(id))
          return
        debug.group(id)
        return utils.transformGroups(code, true)
      },
    })
  }

  plugins.push({
    name: `${NAME}:entry`,

    ...createVirtualModuleLoader({ utils }),
  })

  if (userOptions.transformCSS !== false) {
    plugins.push({
      name: `${NAME}:css`,
      async transform(code, id) {
        await utils.ensureInit()
        if (!utils.isCssTransformTarget(id))
          return
        debug.css(id)
        return utils.transformCSS(code, id)
      },
    })
  }

  return plugins
}

export * from '@windicss/plugin-utils'
export default WindiCssRollupPlugin
