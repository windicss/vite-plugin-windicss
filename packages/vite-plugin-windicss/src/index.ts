import { resolve } from 'node:path'
import type { Plugin, ResolvedConfig, ViteDevServer } from 'vite'
import _debug, { log } from 'debug'
import type { UserOptions, WindiPluginUtils, WindiPluginUtilsOptions } from '@windicss/plugin-utils'
import { createUtils } from '@windicss/plugin-utils'
import { MODULE_ID_VIRTUAL_PREFIX, createVirtualModuleLoader } from '../../shared/virtual-module'
import { createDevtoolsPlugin } from './devtools'
import { NAME } from './constants'
import { getCssModules, reloadChangedCssModules } from './modules'

const debug = {
  hmr: _debug(`${NAME}:hmr`),
  css: _debug(`${NAME}:transform:css`),
  group: _debug(`${NAME}:transform:group`),
  alias: _debug(`${NAME}:transform:alias`),
  memory: _debug(`${NAME}:memory`),
}

function VitePluginWindicss(userOptions: UserOptions = {}, utilsOptions: WindiPluginUtilsOptions = {}): Plugin[] {
  let utils: WindiPluginUtils
  // let viteConfig: ResolvedConfig
  let server: ViteDevServer | undefined
  let viteConfig: ResolvedConfig

  const plugins: Plugin[] = []

  // transform alias
  plugins.push({
    name: `${NAME}:alias`,
    enforce: 'pre',
    configResolved(_config) {
      viteConfig = _config
    },
    async transform(code, id) {
      await utils.ensureInit()
      if (!utils.isDetectTarget(id))
        return
      debug.alias(id)
      return utils.transformAlias(code, !!viteConfig.build.sourcemap)
    },
  })

  // Utilities grouping transform
  if (userOptions.transformGroups !== false) {
    plugins.push({
      name: `${NAME}:groups`,
      enforce: 'pre',
      async transform(code, id) {
        await utils.ensureInit()
        if (!utils.isDetectTarget(id))
          return
        debug.group(id)
        return utils.transformGroups(code, !!viteConfig.build.sourcemap)
      },
    })
  }

  // exposing api
  plugins.push({
    name: NAME,
    get api() {
      return utils
    },
  })

  // CSS Entry via virtual module
  plugins.push({
    name: `${NAME}:entry`,
    enforce: 'post',

    configureServer(_server) {
      server = _server
    },

    async configResolved(_config) {
      // viteConfig = _config
      utils = utilsOptions.utils ?? createUtils(userOptions, {
        name: NAME,
        root: _config.root,
        onConfigurationError(e) {
          if (_config.command === 'build') {
            throw e
          }
          else {
            console.error(`[${NAME}] Error on loading configurations`)
            console.error(e)
          }
        },
        ...utilsOptions,
      })
      await utils.ensureInit()
    },

    ...createVirtualModuleLoader({
      get utils() { return utils },
      get inHmr() {
        return Boolean(server)
      },
    }) as Omit<Plugin, 'name'>,
  })

  let _cssReloadTask: any
  function reloadCssModules(server: ViteDevServer) {
    clearTimeout(_cssReloadTask)
    _cssReloadTask = setTimeout(() => {
      reloadChangedCssModules(server, utils)
    }, 1)
  }

  // HMR
  plugins.push({
    name: `${NAME}:hmr`,
    apply: 'serve',
    enforce: 'pre',

    async configureServer(_server) {
      server = _server

      await utils.ensureInit()
      if (utils.configFilePath)
        server.watcher.add(utils.configFilePath)

      // NOTE: Track changes to the files so that they are re-scanned as needed.
      // Added files are only detected if the user explicitly enables globbing.
      const supportsGlobs = server.config.server.watch?.disableGlobbing === false
      server.watcher.add(supportsGlobs ? utils.globs : await utils.getFiles())
    },

    async handleHotUpdate({ server, file, read }) {
      // resolve normalized file path to system path
      if (resolve(file) === utils.configFilePath) {
        debug.hmr(`config file changed: ${file}`)
        await utils.init()
        setTimeout(() => {
          log('configure file changed, reloading')
          server.ws.send({ type: 'full-reload' })
        }, 0)
        return getCssModules(server)
      }

      if (!utils.isDetectTarget(file))
        return

      utils.extractFile(await read(), file, true)
        .then((changed) => {
          if (changed) {
            debug.hmr(`refreshed by ${file}`)
            reloadCssModules(server)
          }
        })
    },
  })

  const { transformCSS: transformCSSOptions = true } = userOptions

  const transformCSS = (code: string, id: string) => utils.transformCSS(code, id, {
    onLayerUpdated() {
      if (server)
        reloadCssModules(server)
    },
  })

  // CSS transform
  if (transformCSSOptions === true) {
    plugins.push({
      name: `${NAME}:css`,
      async transform(code, id) {
        await utils.ensureInit()
        if (!utils.isCssTransformTarget(id) || id.startsWith(MODULE_ID_VIRTUAL_PREFIX))
          return
        debug.css(id)
        code = transformCSS(code, id)
        if (viteConfig.build.sourcemap) {
          return {
            code: transformCSS(code, id),
            map: { mappings: '' },
          }
        }
        else {
          return code
        }
      },
    })
  }
  else if (typeof transformCSSOptions === 'string') {
    plugins.push({
      name: `${NAME}:css`,
      enforce: transformCSSOptions,
      transform(code, id) {
        if (!utils.isCssTransformTarget(id) || id.startsWith(MODULE_ID_VIRTUAL_PREFIX))
          return
        debug.css(id, transformCSSOptions)
        code = transformCSS(code, id)
        if (viteConfig.build.sourcemap) {
          return {
            code: transformCSS(code, id),
            map: { mappings: '' },
          }
        }
        else {
          return code
        }
      },
    })
  }

  plugins.push({
    name: `${NAME}:css:svelte`,
    api: {
      sveltePreprocess: {
        style({ content, id }: { content: string; id: string }) {
          return {
            code: transformCSS(content, id),
          }
        },
      },
    },
  })

  plugins.push(...createDevtoolsPlugin({
    get utils() { return utils },
  }))

  return plugins
}

export * from '@windicss/plugin-utils'
export default VitePluginWindicss
