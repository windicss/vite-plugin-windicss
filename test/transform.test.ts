import { describe, expect, it } from 'vitest'
import { createUtils } from '../packages/plugin-utils/src'

describe('transfrom', () => {
  it('groups', async () => {
    const cases = [
      'bg-white font-light sm:hover:(bg-gray-100 font-medium)',
      '-sm:hover:(p-1 p-2)',
      '+sm:(p-1 p-2)',
      'dark:+lg:(p-1 p-2)',
      '@lg:(p-1 p-2)',
      'md:(w-40vw pr-4.5rem)',
      '<md:(grid grid-cols-[1fr,50%])',
      '!hover:(m-2 p-2)',
      'hover:(!m-2 p-2)',
      'md:(w-1/2 h-[calc(100%-4rem)])',
      'hover:(\n!m-2 \np-2\n)',
    ]
    const utils = createUtils()

    for (const c of cases) {
      const transformed = utils.transformGroups(c)?.code
      expect(transformed).toMatchSnapshot(`"${c}"`)
    }
  })

  it('css directives', async () => {
    const utils = createUtils({
      preflight: false,
      scan: false,
    })

    await utils.init()

    expect(await utils.generateCSS()).toMatchSnapshot()

    expect(utils.transformCSS(`
      .rounded-box {
        border-radius: var(--rounded-box, 1rem);
      }
      .card {
        color: black;
        @apply rounded-box shadow hover:shadow-xl;
      }
      .artboard {
        @apply rounded-box;
      }
    `, 'group0')).toMatchSnapshot('basic @apply')

    expect(await utils.generateCSS()).toMatchSnapshot()

    expect(utils.transformCSS(`
      .btn {
        padding: 5px;
      }
      @layer utilities {
        .btn-utilities {
          @apply text-red-200;
          font-size: 2rem;
        }
      }
      @layer base {
        .btn-base {
          @apply text-red-300;
          font-size: 1.5rem;
        }
      }
      @layer components {
        .btn-components {
          @apply text-red-400;
          font-size: 1rem;
        }
      }
    `, 'group1')).toMatchSnapshot('basic @layer')

    // should merge with utilities
    utils.addClasses(['p-4'])

    expect(await utils.generateCSS()).toMatchSnapshot('@layer all')
    expect(await utils.generateCSS('base')).toMatchSnapshot('@layer base')
    expect(await utils.generateCSS('components')).toMatchSnapshot('@layer components')
    expect(await utils.generateCSS('utilities')).toMatchSnapshot('@layer utilities')

    // should replace previous state
    expect(utils.transformCSS(`
      @layer components {
        .btn-components {
          @apply text-red-500;
          font-size: 1.5rem;
        }
      }
    `, 'group1')).toMatchSnapshot()

    const base = await utils.generateCSS('base')
    const components = await utils.generateCSS('components')
    expect(base).not.toContain('.btn-base')
    expect(components).toMatchSnapshot('@layer components')
    expect(await utils.generateCSS('utilities')).toMatchSnapshot('@layer utilities')
  })
  it('css windi keyframes', async () => {
    const utils = createUtils({
      preflight: false,
      scan: false,
    })

    await utils.init()

    expect(await utils.generateCSS()).toMatchSnapshot()

    const css = `
      .pulse-class {
         @apply relative w-40 h-40 rounded-full bg-teal-600 opacity-80 animate-pulse;
      }
    `

    await utils.extractFile(css, 'group0', true)

    const transformed = utils.transformCSS(css, 'group0', { globaliseKeyframes: true })

    expect(transformed).not.toContain('@keyframes')
    expect(transformed).toMatchSnapshot('keyframes')

    const moduleCSS = await utils.generateCSS()
    expect(moduleCSS).toContain('@keyframes')
    expect(moduleCSS).toMatchSnapshot('keyframes module')
  })
  it('css custom keyframes', async () => {
    const utils = createUtils({
      preflight: false,
      scan: false,
    })

    await utils.init()

    expect(await utils.generateCSS()).toMatchSnapshot()

    const css = `
      .test {
        animation: my-animation 2s infinite linear;
      }
      @keyframes my-animation {
        100% {
          -webkit-transform: rotate(360deg);
        }
      }
    `

    await utils.extractFile(css, 'group0', true)

    const transformed = utils.transformCSS(css, 'group0', { globaliseKeyframes: true })

    expect(transformed).toContain('@keyframes')
    expect(transformed).toMatchSnapshot('keyframes')

    const moduleCSS = await utils.generateCSS()
    expect(moduleCSS).not.toContain('@keyframes')
    expect(moduleCSS).toMatchSnapshot('keyframes module')
  })
})
