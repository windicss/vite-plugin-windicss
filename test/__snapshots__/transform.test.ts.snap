// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transfrom css custom keyframes 1`] = `
"/* windicss layer base */

/* windicss layer components */

/* windicss layer utilities */"
`;

exports[`transfrom css custom keyframes: keyframes 1`] = `
".test {
  animation: my-animation 2s infinite linear;
}
@keyframes my-animation {
  100% {
    -webkit-transform: rotate(360deg);
  }
}"
`;

exports[`transfrom css custom keyframes: keyframes module 1`] = `
"/* windicss layer base */

/* windicss layer components */

/* windicss layer utilities */"
`;

exports[`transfrom css directives 1`] = `
"/* windicss layer base */

/* windicss layer components */

/* windicss layer utilities */"
`;

exports[`transfrom css directives 2`] = `
"/* windicss layer base */

/* windicss layer components */

/* windicss layer utilities */"
`;

exports[`transfrom css directives 3`] = `""`;

exports[`transfrom css directives: @layer all 1`] = `
"/* windicss layer base */
.btn-base {
  --tw-text-opacity: 1;
  color: rgba(252, 165, 165, var(--tw-text-opacity));
  font-size: 1.5rem;
}
/* windicss layer components */
.btn-components {
  --tw-text-opacity: 1;
  color: rgba(248, 113, 113, var(--tw-text-opacity));
  font-size: 1rem;
}
/* windicss layer utilities */
.p-4 {
  padding: 1rem;
}
.btn-utilities {
  --tw-text-opacity: 1;
  color: rgba(254, 202, 202, var(--tw-text-opacity));
  font-size: 2rem;
}"
`;

exports[`transfrom css directives: @layer base 1`] = `
"/* windicss layer base */
.btn-base {
  --tw-text-opacity: 1;
  color: rgba(252, 165, 165, var(--tw-text-opacity));
  font-size: 1.5rem;
}"
`;

exports[`transfrom css directives: @layer components 1`] = `
"/* windicss layer components */
.btn-components {
  --tw-text-opacity: 1;
  color: rgba(248, 113, 113, var(--tw-text-opacity));
  font-size: 1rem;
}"
`;

exports[`transfrom css directives: @layer components 2`] = `
"/* windicss layer components */
.btn-components {
  --tw-text-opacity: 1;
  color: rgba(239, 68, 68, var(--tw-text-opacity));
  font-size: 1.5rem;
}"
`;

exports[`transfrom css directives: @layer utilities 1`] = `
"/* windicss layer utilities */
.p-4 {
  padding: 1rem;
}
.btn-utilities {
  --tw-text-opacity: 1;
  color: rgba(254, 202, 202, var(--tw-text-opacity));
  font-size: 2rem;
}"
`;

exports[`transfrom css directives: @layer utilities 2`] = `
"/* windicss layer utilities */
.p-4 {
  padding: 1rem;
}"
`;

exports[`transfrom css directives: basic @apply 1`] = `
".rounded-box {
  border-radius: var(--rounded-box, 1rem);
}
.card {
  color: black;
  --tw-shadow: 0 1px 3px 0 rgb(0 0 0/0.1),0 1px 2px -1px rgb(0 0 0/0.1);
  --tw-shadow-colored: 0 1px 3px 0 var(--tw-shadow-color),0 1px 2px -1px var(--tw-shadow-color);
  -webkit-box-shadow: var(--tw-ring-offset-shadow,0 0 #0000),var(--tw-ring-shadow,0 0 #0000),var(--tw-shadow);
  box-shadow: var(--tw-ring-offset-shadow,0 0 #0000),var(--tw-ring-shadow,0 0 #0000),var(--tw-shadow);
  border-radius: var(--rounded-box, 1rem);
}
.card:hover {
  --tw-shadow: 0 20px 25px -5px rgb(0 0 0/0.1),0 8px 10px -6px rgb(0 0 0/0.1);
  --tw-shadow-colored: 0 20px 25px -5px var(--tw-shadow-color),0 8px 10px -6px var(--tw-shadow-color);
  -webkit-box-shadow: var(--tw-ring-offset-shadow,0 0 #0000),var(--tw-ring-shadow,0 0 #0000),var(--tw-shadow);
  box-shadow: var(--tw-ring-offset-shadow,0 0 #0000),var(--tw-ring-shadow,0 0 #0000),var(--tw-shadow);
}
.artboard {
  border-radius: var(--rounded-box, 1rem);
}"
`;

exports[`transfrom css directives: basic @layer 1`] = `
".btn {
  padding: 5px;
}"
`;

exports[`transfrom css windi keyframes 1`] = `
"/* windicss layer base */

/* windicss layer components */

/* windicss layer utilities */"
`;

exports[`transfrom css windi keyframes: keyframes 1`] = `
".pulse-class {
  --tw-bg-opacity: 1;
  background-color: rgba(13, 148, 136, var(--tw-bg-opacity));
  border-radius: 9999px;
  height: 10rem;
  opacity: 0.8;
  position: relative;
  width: 10rem;
  -webkit-animation: pulse 2s cubic-bezier(0.4, 0, 0.6, 1) infinite;
  animation: pulse 2s cubic-bezier(0.4, 0, 0.6, 1) infinite;
}"
`;

exports[`transfrom css windi keyframes: keyframes module 1`] = `
"/* windicss layer base */

/* windicss layer components */

/* windicss layer utilities */
.bg-teal-600 {
  --tw-bg-opacity: 1;
  background-color: rgba(13, 148, 136, var(--tw-bg-opacity));
}
.rounded-full {
  border-radius: 9999px;
}
.h-40 {
  height: 10rem;
}
.opacity-80 {
  opacity: 0.8;
}
.relative {
  position: relative;
}
.w-40 {
  width: 10rem;
}
@keyframes pulse {
  .pulse-class {
    opacity: 1;
    opacity: .5;
  }
}
@-webkit-keyframes pulse {
  .pulse-class {
    opacity: 1;
    opacity: .5;
  }
}"
`;

exports[`transfrom groups: "!hover:(m-2 p-2)" 1`] = `"!hover:m-2 !hover:p-2"`;

exports[`transfrom groups: "+sm:(p-1 p-2)" 1`] = `"+sm:p-1 +sm:p-2"`;

exports[`transfrom groups: "<md:(grid grid-cols-[1fr,50%])" 1`] = `"<md:grid <md:grid-cols-[1fr,50%]"`;

exports[`transfrom groups: "@lg:(p-1 p-2)" 1`] = `"@lg:p-1 @lg:p-2"`;

exports[`transfrom groups: "-sm:hover:(p-1 p-2)" 1`] = `"-sm:hover:p-1 -sm:hover:p-2"`;

exports[`transfrom groups: "bg-white font-light sm:hover:(bg-gray-100 font-medium)" 1`] = `"bg-white font-light sm:hover:bg-gray-100 sm:hover:font-medium"`;

exports[`transfrom groups: "dark:+lg:(p-1 p-2)" 1`] = `"dark:+lg:p-1 dark:+lg:p-2"`;

exports[`transfrom groups: "hover:(
!m-2 
p-2
)" 1`] = `"!hover:m-2 hover:p-2"`;

exports[`transfrom groups: "hover:(!m-2 p-2)" 1`] = `"!hover:m-2 hover:p-2"`;

exports[`transfrom groups: "md:(w-1/2 h-[calc(100%-4rem)])" 1`] = `"md:w-1/2 md:h-[calc(100%-4rem)]"`;

exports[`transfrom groups: "md:(w-40vw pr-4.5rem)" 1`] = `"md:w-40vw md:pr-4.5rem"`;
